### PR TITLE
CORE-5001 Consensual ledger structures

### PR DESCRIPTION
**Pull request by vlajos**
ledger-common:
* PrivacySaltImpl
* WireTransaction
* TransactionMetaData/CpkIdentifier

ledger-consensual:
* ConsensualLedgerServiceImpl
* PartyImpl
* ConsensualLedgerTransactionImpl
* ConsensualSignedTransactionImpl
* ConsensualTransactionBuilderImpl

Status:
* Unit/Integration tests, corDapp for checking ConsensualLedgerServiceI.getTranssactionBuilder() pass
* Unit test, corDapp for building ConsensualSignedTransactionImpl with the above pass. (Integration test won't be available for this since that needs serialization whichs needs a running fiber which is not fully supported in integration tests at the moment.)

To do:
* Doc
* More tests
* Review Detekt...
* ~~Split serialization specific changes into dedicated PR~~
* ~~Split test-serialization module into dedicated PR (maybe dedup testSerializationContexts?)~~

API: https://github.com/corda/corda-api/pull/487

### CODE REVIEW

File: libs/ledger/ledger-common-impl/src/test/java/net/corda/ledger/common/impl/transactions/PrivacySaltJavaApiTest.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package net.corda.ledger.common.impl.transactions;
```
Comment (blsemo) [inline libs/ledger/ledger-common-impl/src/test/java/net/corda/ledger/common/impl/transactions/PrivacySaltJavaApiTest.java]: I am not sure we need a JavaAPI test on an implementation class?

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/PartyImpl.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package net.corda.ledger.consensual.impl
+
+import net.corda.v5.base.types.MemberX500Name
+import net.corda.v5.ledger.consensual.Party
+import java.security.PublicKey
+
+class PartyImpl(override val name: MemberX500Name, override val owningKey: PublicKey) : Party {
```
Comment (blsemo) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/PartyImpl.kt:11]: There are a couple of questions around party:
- Shouldn't that be a common type?
- Would we need a Party interface that doesn't have an X500 name so we can have e.g. anonymous parties or account parties?

File: libs/ledger/ledger-common-impl/src/test/java/net/corda/ledger/common/impl/transactions/PrivacySaltJavaApiTest.java (hunk 1 above)
Comment (vlajos) [reply libs/ledger/ledger-common-impl/src/test/java/net/corda/ledger/common/impl/transactions/PrivacySaltJavaApiTest.java]: Thanks. I have removed it.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/PartyImpl.kt (hunk 2 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/PartyImpl.kt:11]: I have moved this comment here: https://r3-cev.atlassian.net/browse/CORE-5936

File: libs/crypto/crypto-serialization-impl/src/main/java/net/corda/crypto/impl/serialization/package-info.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+@Export
+package net.corda.crypto.impl.serialization;
+
+import org.osgi.annotation.bundle.Export;
```
Comment (blsemo) [inline libs/crypto/crypto-serialization-impl/src/main/java/net/corda/crypto/impl/serialization/package-info.java:4]: Do we need an export for this impl package?

File: libs/ledger/ledger-common-impl/src/main/java/net/corda/ledger/common/impl/transaction/package-info.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+@Export
```
Comment (blsemo) [inline libs/ledger/ledger-common-impl/src/main/java/net/corda/ledger/common/impl/transaction/package-info.java:1]: Do we need a package export for this impl package?

File: libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/WireTransaction.kt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+internal val COMPONENT_MERKLE_TREE_ENTROPY_ALGORITHM_NAME = DigestAlgorithmName.SHA2_256D
+
+const val ALL_LEDGER_METADATA_COMPONENT_GROUP_ID = 0
+
+@CordaSerializable
+class WireTransaction(
+    private val merkleTreeFactory: MerkleTreeFactory,
+    private val digestService: DigestService,
+    val privacySalt: PrivacySalt,
+    val componentGroupLists: List<List<ByteArray>>
```
Comment (blsemo) [inline libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/WireTransaction.kt:34]: Do we have a custom serializer for this class to ensure we don't serialize all these services?

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import net.corda.ledger.common.impl.transaction.WireTransaction
+import net.corda.v5.application.crypto.DigitalSignatureAndMetadata
+import net.corda.v5.application.serialization.SerializationService
+import net.corda.v5.crypto.SecureHash
+import net.corda.v5.crypto.isFulfilledBy
+import net.corda.v5.ledger.consensual.transaction.ConsensualLedgerTransaction
+import net.corda.v5.ledger.consensual.transaction.ConsensualSignedTransaction
+import java.security.PublicKey
+
+class ConsensualSignedTransactionImpl(
```
Comment (blsemo) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt:12]: How do we serialize this? Or do we not serialize it but store the wire TX and the signatures separately?

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        if (this === other) return true
+        if (other !is ConsensualTransactionBuilderImpl) return false
+        if (other.states.size != states.size) return false
+
+        return other.states.withIndex().all{
+            it.value == states[it.index]
+        }
+    }
+
+    override fun hashCode(): Int = states.hashCode()
```
Comment (blsemo) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:42]: Is that good enough? So two TxBuilders with empty set of states would end up with the same hash even though they are are different?

Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:42]: There is nothing else in the TxBuilders, only the states. Everything else gets populated automatically.

File: libs/ledger/ledger-common-impl/src/main/java/net/corda/ledger/common/impl/transaction/package-info.java (hunk 4 above)
Comment (vlajos) [reply libs/ledger/ledger-common-impl/src/main/java/net/corda/ledger/common/impl/transaction/package-info.java:1]: Without this export, integration tests fail with this:
```
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=net.corda.ledger-consensual-impl)
          ⇒ [net.corda.ledger-consensual-impl version=5.0.0.0-SNAPSHOT]
              ⇒ osgi.wiring.package: (osgi.wiring.package=net.corda.ledger.common.impl.transaction)
```

File: libs/crypto/crypto-serialization-impl/src/main/java/net/corda/crypto/impl/serialization/package-info.java (hunk 3 above)
Comment (vlajos) [reply libs/crypto/crypto-serialization-impl/src/main/java/net/corda/crypto/impl/serialization/package-info.java:4]: Without this export, integration tests fail with this:
```
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=ledger-consensual-impl-tests)
          ⇒ [ledger-consensual-impl-tests version=5.0.0.0-SNAPSHOT]
              ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.internal.serialization.amqp.helper)(version>=5.0.0)(!(version>=6.0.0)))
                  ⇒ [net.corda.test-serialization version=5.0.0.0-SNAPSHOT]
                      ⇒ osgi.wiring.package: (osgi.wiring.package=net.corda.crypto.impl.serialization)
```

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt (hunk 6 above)
Comment (blsemo) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt:12]: Separate JIRA - let's refactor later: https://r3-cev.atlassian.net/browse/CORE-5114

**Review by mnesbit — APPROVED** [review]
OK

**Pull request merged**